src/cdk/drag-drop/drop-list-ref.ts

```diff
@@ -820,7 +820,101 @@ export class DropListRef<T = any> {
    * @param y Position of the item along the Y axis.
    */
   _getSiblingContainerFromPosition(item: DragRef, x: number, y: number): DropListRef | undefined {
-    return this._siblings.find(sibling => sibling._canReceive(item, x, y));
+    // Possible targets include siblings and 'this'
+    let targets = [this, ...this._siblings];
+
+    // Only consider targets where the drag postition is within the client rect
+    // (this avoids calling enterPredicate on each possible target)
+    let matchingTargets = targets.filter(ref => {
+      return ref._clientRect && isInsideClientRect(ref._clientRect, x, y);
+    });
+
+    // Stop if no targets match the coordinates
+    if (matchingTargets.length == 0) {
+        return undefined;
+    }
+
+    // Order candidates by DOM hierarchy and z-index
+    let orderedMatchingTargets = this._orderByHierarchy(matchingTargets);
+
+    // The drop target is the last matching target in the list
+    let matchingTarget = orderedMatchingTargets[orderedMatchingTargets.length - 1];
+
+    // Only return matching target if it is a sibling
+    if (matchingTarget === this) {
+      return undefined;
+    }
+
+    // Can the matching target receive the item?
+    if (!matchingTarget._canReceive(item, x, y)) {
+      return undefined;
+    }
+
+    // Return matching target
+    return matchingTarget;
+  }
+
+  /**
+   * Sort a list of DropListRef in such that forevery nested pair drop containers, the
+   * outer drop container appear before the inner drop container.
+   * @param refs List of DropListRefs.
+   */
+  private _orderByHierarchy(refs: DropListRef[]): DropListRef[] {
```

Comment:
This whole method is a bit concerning in terms of performance, because we call it for every pixel that the user has dragged. The DOM actually has an API that returns the relative position between two nodes: https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition. That being said, I don't know what the performance implications of the native API are either.

Alternatively, we could cache the tree structure after the first check, because the DOM is unlikely to change while the user is dragging.

Reply:
I share your concerns, although I might not become a problem in practice. The number of references ordred in each call to `_orderByHierarchy` is limited by the depth of the drop container hierarchy. Hence, if the hierarchy is 10 levels deep (a rather massive tree), the method will be called with a list of at most 10 drop containers to sort. This is not a huge effort for a modern computer / device.

Anyway, I think caching would be a good idea.

It might also make sense to re-implement the function using "compareDocumentPosition" rather than traversing the hierarchy manually. 

I'll have a go at it.